Applications/Utils/ModelPreparation/PartitionMesh/MeshPartitioning.cpp

```diff
@@ -0,0 +1,493 @@
+        for (PetscInt i=0; i<num_non_ghost_nodes; i++)
+        {
+            const unsigned node_id = partition_nodes[ i +  partition_start_node_id[ipart] ]->getID();
+
+            // Search the elements connected to this nodes
+            std::vector<Element*> const& connected_elems = _nodes[node_id]->getElements();
+
+            for (const auto elem : connected_elems)
+            {
+                // If checked
+                if (elems_reseved[elem->getID()])
+                    continue;
+
+                std::vector<unsigned> nonghost_nodes_local_ids;
+                unsigned ghost_counter = 0;
+                for (unsigned kk=0; kk<elem->getNNodes(); kk++)
+                {
+                    if (nodes_reseved[elem->getNodeIndex(kk)])
+                    {
+                        nonghost_nodes_local_ids.push_back(kk);
+                    }
+                    else
+                    {
+                        ghost_counter++;
+                    }
+                }
+
+                // All nodes of this element are inside this partition
+                if (ghost_counter == 0)
+                {
+                    partition_regular_elements.push_back(elem);
+                }
+                else // ghost element
+                {
+                    partition_ghost_elements.push_back(elem);
+                    elems_non_ghost_nodes_local_ids[elem->getID()]
+                        = std::move(nonghost_nodes_local_ids);
+                }
+
+                elems_reseved[elem->getID()] = true;
+            }
+        }
+
+        // Add ghost nodes in ghost elements to the node vector of this partition
+        // nodes_reseved is reused without cleaning
+        // Mark the non-ghost nodes for each ghost element
+        for (const auto ghost_elem : partition_ghost_elements)
+        {
+            for (unsigned k=0; k<ghost_elem->getNNodes(); k++)
+                nodes_reseved[ghost_elem->getNodeIndex(k)] = false;
+
+            // Mark non-ghost nodes
+            std::vector<unsigned>& local_node_ids = elems_non_ghost_nodes_local_ids[ghost_elem->getID()];
+            for (std::size_t k=0; k<local_node_ids.size(); k++)
+                nodes_reseved[ghost_elem->getNodeIndex(local_node_ids[k])] = true;
+        }
+        // Add the ghost nodes to the node vector of this partition
+        for (const auto ghost_elem : partition_ghost_elements)
+        {
+            for (unsigned k=0; k<ghost_elem->getNNodes(); k++)
+            {
+                const unsigned node_id = ghost_elem->getNodeIndex(k);
+                if (nodes_reseved[node_id])
+                    continue;
+                nodes_reseved[node_id] = true;
+                partition_nodes.push_back(ghost_elem->getNode(k));
+            }
+        }
+
+        partition_end_node_id[ipart] = partition_nodes.size();
+        // Renumber
+        PetscInt local_id = 0;
+
+        std::vector<unsigned> nodes_local_ids_partition;
```

Comment:
Maybe better: `vector<std::size_t>`. As @endJunction noted above: Can be filled via constructor.